jena-arq/src/main/java/org/apache/jena/riot/tokens/TokenizerText.java

```diff
@@ -439,37 +458,42 @@ private Token parseToken() {
 
         */
 
-        // TODO extract readNumberNoSign
-
-        int signCh = 0;
-
         if ( ch == CH_PLUS || ch == CH_MINUS ) {
             reader.readChar();
             int ch2 = reader.peekChar();
-
-            if ( !range(ch2, '0', '9') ) {
-                // ch was end of symbol.
-                // reader.readChar();
+            if ( !range(ch2, '0', '9') && ch2 != CH_DOT ) {
+                // Not a number.
                 if ( ch == CH_PLUS )
                     token.setType(TokenType.PLUS);
                 else
                     token.setType(TokenType.MINUS);
                 return token;
             }
-
-            // Already got a + or - ...
-            // readNumberNoSign
-            // Because next, old code processes signs.
-            reader.pushbackChar(ch);
-            signCh = ch;
-            // Drop to next "if"
+            // ch2 not consumed.
+            boolean charactersConsumed = readNumber(ch, false);
+            if ( ! charactersConsumed ) {
+                if ( ch == CH_PLUS )
+                    token.setType(TokenType.PLUS);
+                else
+                    token.setType(TokenType.MINUS);
+            }
+            return token;
         }
 
-        if ( ch == CH_PLUS || ch == CH_MINUS || range(ch, '0', '9') ) {
-            // readNumberNoSign
-            readNumber();
-            if ( Checking )
-                checkNumber(token.getImage(), token.getImage2());
+        if ( range(ch, '0', '9')  ) {
```

Comment:
So no need to check the plus and minus here as it was handled above... but do we need to handle the `CH_DOT` here too? I'm guessing the previous `if` handles `-0`, `+1.2`, etc., and this one handles `0`, `1.`?

Reply:
Leading DOT is handled in the `case DOT` rule at line 330. It works by pushing back the DOT into the read stream.
It means there are 3 routes to `readNumber` spread out across the token loop.

Reply:
Got it, had missed that branch!